src/modules/did/did.service.ts

```diff
@@ -495,4 +517,71 @@ export class DIDService implements OnModuleInit, OnModuleDestroy {
       this.logger.debug(inspect(jobsCounts, { depth: 2, colors: true }));
     }
   }
+
+  private async getChangedIdentities(
+    fromBlock: number,
+    topBlock: number
+  ): Promise<string[]> {
+    const didEventFilters = [
+      this.didRegistry.filters.DIDAttributeChanged(null),
+      this.didRegistry.filters.DIDDelegateChanged(null),
+      this.didRegistry.filters.DIDOwnerChanged(null),
+    ];
+    const events = [];
+    while (fromBlock < topBlock) {
+      const toBlock = Math.min(
+        topBlock,
+        fromBlock + this.MAX_EVENTS_QUERY_INTERVAL
+      );
+
+      const intervalEvents = (
+        await Promise.all(
+          didEventFilters.map((filter) =>
+            this.didRegistry.queryFilter(filter, fromBlock, toBlock)
+          )
+        )
+      ).flat();
+      this.logger.debug(
+        `Fetched ${intervalEvents.length} DID events from interval [${fromBlock}, ${toBlock}]`
+      );
+      events.push(...intervalEvents);
+      fromBlock += this.MAX_EVENTS_QUERY_INTERVAL;
+    }
+    this.logger.debug(`Update document events count ${events.length}`);
+    const changedIdentities = events.map((event) => {
+      return event.args.identity;
+    });
+    // Deduplicate identities if they are appears in multiple events
+    return [...new Set(changedIdentities).values()];
+  }
+
+  private async markStaleDocuments() {
+    const syncs = await this.latestDidSyncRepository.find({
+      order: { createdDate: 'DESC' },
+    });
+    const fromBlock = syncs.length > 0 ? syncs[0].block : 0;
+    const topBlock = await this.provider.getBlockNumber();
+    this.logger.debug(
+      `Getting DID update events from block ${fromBlock} to block ${topBlock}...`
+    );
+    const changedIdentities = await this.getChangedIdentities(
+      fromBlock,
+      topBlock
+    );
+    const staleDIDs = (
+      await this.didRepository.find({ select: ['id'] })
+    ).filter((doc) => {
+      const identity = addressOf(doc.id);
+      return changedIdentities.includes(identity);
+    });
+    await this.didSyncStatusRepository
+      .createQueryBuilder()
+      .useTransaction(true)
+      .update(DidSyncStatusEntity)
+      .set({ status: DidSyncStatus.Stale })
+      .where({ document: In(staleDIDs) })
+      .execute();
```

Comment:
Would this query work for existing cached documents (already existing in the cache I mean)? Maybe I missed it, but I don't see a migration that adds a `DidSyncStatusEntity` for all existing cached DID Documents.

Reply:
Yes, I missed that